### PR TITLE
Initial implementation of first stage of DOM  Isolation

### PR DESCRIPTION
**Pull request by cromwellian**
Limitations:
* handles custom elements with a whitelist
* handles custom attributes work a hack auto-whitelist
* Doesn't deal with unsafe-html attribute or sub-templates

TODO:
* Handle custom attributes via observeAttributes() or declaragtive syntax
* Deal with sub-templates
* sanitize unsafe-html
* CSP policy to block unsafe evaluations

---
This change is [Reviewable](https://reviewable.io/reviews/polymerlabs/arcs/1790)

### CODE REVIEW

**Comment by cromwellian** [comment]
I'm sending this without a test, I'm running into trouble with transitive node_modules not loading correctly in unit tests, so rather than delay this for review, I've decided to send it early while I debug this loader issue.

File: runtime/ts/isolate/sanitize.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    attribute?: Node;
+}
+
+interface Node {
+    name: string;
+}
+
+export function sanitize(unsafeString: string): string {
+
+    // TODO(cromwellian): investigate adding a WeakMap memoize here? Ideally, templates should only be purified once.
```
Comment (sjmiles) [inline runtime/ts/isolate/sanitize.ts]: Fwiw, templates also undergo non-trivial processing by Xen, so there is incentive to memoize at a higher level. IOW, once a (String) template has been sanitized, converted to DOM Template, and pre-processed by Xen, the resulting DOM Template should be cached and re-used.

File: runtime/ts/isolate/sanitize.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ */
+const ESCAPED_STYLE_PCT = "__style__pct__";
+
+/*
+ * DOMPurify must be told about non-standard HTML elements (e.g. web components). For now we just whitelist them here,
+ * but we should consider having some mechanism in the runtime to auto-discover these. Note, if you add a new
+ * Web Component to the shell, it must be registered here.
+ */
+const CUSTOM_ELEMENTS: string[] = [
+    'app-shell',
```
Comment (sjmiles) [inline runtime/ts/isolate/sanitize.ts]: None of the Shell elements should be allowed by Particles. There is a much shorter list of elements specifically whitelisted for use in Particle DOM. Currently, these are collated for import [here](https://github.com/PolymerLabs/arcs/blob/master/shell/apps/common/whitelisted.js). Probably we should have this import supply only the paths and make it the source of truth for the sanitizer and a Shell loader.

File: runtime/ts/isolate/sanitize.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    'x-toast',
+    'remove',
+    'xen-explorer'];
+
+/*
+ * Whitelist attribute names not to prune. Ideally, most of these would be auto-discovered somehow such as by tracking
+ * calls to observedAttributes().
+ */
+const CUSTOM_ATTRIBUTES = [
+    ESCAPED_STYLE_PCT
```
Comment (sjmiles) [inline runtime/ts/isolate/sanitize.ts]: If `style%` is problematic we can replace it with some other variation.

**Pull request closed without merging**